### PR TITLE
[easy] Extend function for custom gates

### PR DESCRIPTION
**Pull request by querolita**
This PR provides extend functions to all of Xor, And, Not, Rot.  This function allows the gadgets to appear anywhere in the circuit (not only assuming the row 0).

### CODE REVIEW

File: kimchi/src/tests/foreign_field_add.rs [hunk 1]
```diff
@@ -1453,3 +1456,106 @@ where
 
     (Ok(()), witness)
 }
+
+#[test]
+// Finalization test
+fn test_ffadd_finalization() {
+    // Includes a row to store value 1
+    let num_inputs = 1;
```
Comment (jspada) [inline kimchi/src/tests/foreign_field_add.rs:1464]: Call this `num_public_inputs`

File: kimchi/src/tests/foreign_field_add.rs [hunk 2]
```diff
@@ -1453,3 +1456,106 @@ where
...
+    let operation = &[FFOps::Add];
+    let modulus = BigUint::from_bytes_be(&[
+        0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFE, 0xFF, 0xFF, 0xFF, 0xFF,
+        0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFE, 0xFF, 0xFF,
+        0xFC, 0x2F,
+    ]);
+
+    // circuit
```
Comment (jspada) [inline kimchi/src/tests/foreign_field_add.rs:1472]: You should comment the circuit layout here for reference below when creating copy constraints.

File: kimchi/src/tests/foreign_field_add.rs [hunk 3]
```diff
@@ -1453,3 +1456,106 @@ where
...
+        }
+        // extend range checks for all of left, right, output, and bound
+        let left = (witness[0][1], witness[1][1], witness[2][1]);
+        range_check::witness::extend_multi(&mut witness, left.0, left.1, left.2);
+        let right = (witness[3][1], witness[4][1], witness[5][1]);
+        range_check::witness::extend_multi(&mut witness, right.0, right.1, right.2);
+        let output = (witness[0][2], witness[1][2], witness[2][2]);
+        range_check::witness::extend_multi(&mut witness, output.0, output.1, output.2);
+        let bound = (witness[0][3], witness[1][3], witness[2][3]);
+        range_check::witness::extend_multi(&mut witness, bound.0, bound.1, bound.2);
```
Comment (jspada) [inline kimchi/src/tests/foreign_field_add.rs:1524]: Nice clarity

File: kimchi/src/tests/not.rs [hunk 4]
```diff
@@ -46,6 +42,36 @@ const RNG_SEED: [u8; 32] = [
     89, 29, 13, 250, 215, 172, 130, 24, 164, 162,
 ];
 
+// Creates as many negations as the number of inputs. The inputs must fit in the native field.
+// We start at the row 0 using generic gates to perform the negations.
+// Input: a vector of words to be negated, and the number of bits (all the same)
+// Panics if the bits length is too small for the inputs
```
Comment (jspada) [inline kimchi/src/tests/not.rs:48]: possible not to panic and error instead?

File: kimchi/src/tests/not.rs [hunk 5]
```diff
@@ -46,6 +42,36 @@ const RNG_SEED: [u8; 32] = [
...
+    let mut witness: [Vec<F>; COLUMNS] = array::from_fn(|_| vec![F::zero(); 1]);
+    witness[0][0] = F::from(2u8).pow(&[bits as u64]) - F::one();
+    not::extend_not_witness_unchecked_length(&mut witness, inputs, bits);
+    witness
+}
+
+// Create a Not witness for less than 255 bits (native field) starting at row 0
+// Input: first input and optional bit length
+// If `bits` is not provided, the negation is performed using the length of the `input` in bits.
+// If `bits` is provided, the negation takes the maximum length between `bits` and that of `input`.
```
Comment (jspada) [inline kimchi/src/tests/not.rs:62]: nit: maximum length of `bits` and `input`.

File: kimchi/src/tests/rot.rs [hunk 6]
```diff
@@ -29,24 +36,56 @@ type VestaScalarSponge = DefaultFrSponge<Fp, SpongeParams>;
...
+    word: u64,
+    rot: u32,
+    side: RotMode,
+) -> [Vec<G::ScalarField>; COLUMNS]
+where
+    G::BaseField: PrimeField,
+{
+    // Include the zero row
+    let mut witness: [Vec<G::ScalarField>; COLUMNS] =
+        array::from_fn(|_| vec![G::ScalarField::from(0u32)]);
```
Comment (jspada) [inline kimchi/src/tests/rot.rs:71]: `array::from_fn(|_| vec![G::ScalarField::zero()]);`

File: kimchi/src/tests/rot.rs [hunk 7]
```diff
@@ -259,3 +297,83 @@ fn test_bad_constraints() {
         })
     );
 }
+
+#[test]
+// Finalization test
+fn test_rot_finalization() {
+    // Includes the actual input of the rotation and a row with the zero value
+    let num_inputs = 2;
```
Comment (jspada) [inline kimchi/src/tests/rot.rs:305]: nit: `num_public_inputs`

File: kimchi/src/tests/xor.rs [hunk 8]
```diff
@@ -289,7 +300,55 @@ fn test_bad_xor_decompsition() {
...
 #[test]
-// Test that a 16-bit XOR gate fails if the witness does not correspond to a XOR operation
+// Tests that the extend xor function works as expected
+fn test_extend_xor() {
+    let bits = Some(16);
+    let rng = &mut StdRng::from_seed(RNG_SEED);
+    let input1: PallasField = rng.gen(None, bits);
+    let input2: PallasField = rng.gen(None, bits);
+
+    // If user specified a concrete number of bits, use that (if they are sufficient to hold both inputs)
```
Comment (jspada) [inline kimchi/src/tests/xor.rs:310]: nit: confusing sentence -- there's no user here, just rng :-)

File: kimchi/src/tests/xor.rs [hunk 9]
```diff
@@ -330,3 +390,81 @@ fn test_bad_xor() {
...
+        }
+        gates
+    };
+
+    // witness
+    let witness = {
+        let mut cols: [_; COLUMNS] = array::from_fn(|_col| vec![Fp::zero(); num_inputs]);
+
+        // initialize the 2 inputs
+        let input1 = 0xDC811727DAF22EC15927D6AA275F406Bu128;
```
Comment (jspada) [inline kimchi/src/tests/xor.rs:428]: Can you just do `0xDC811727DAF22EC15927D6AA275F406Bu128.into()` ?

File: kimchi/src/tests/xor.rs [hunk 10]
```diff
@@ -330,3 +390,81 @@ fn test_bad_xor() {
...
+        let input2 = 0xA4F4417AF072DF9016A1EAB458DA80D1u128;
+        cols[0][0] = input1.into();
+        cols[0][1] = input2.into();
+
+        xor::extend_xor_witness::<Fp>(&mut cols, input1.into(), input2.into(), 128);
+        cols
+    };
+
+    let index = {
```
Comment (jspada) [inline kimchi/src/tests/xor.rs:437]: This thing feels very repetitive and could be refactored into a function to save some lines of code.  However, not for now.  Maybe after hard-fork in another PR.

**Review by jspada — APPROVED** [review]
Nice improvement on the interface, the extend functions will be much more practical to use!

Checked it as closely as possible.

Approved with some feedback, suggestions and nits.

File: kimchi/src/circuits/polynomials/rot.rs [hunk 11]
```diff
@@ -282,6 +293,13 @@ fn init_rot64<F: PrimeField>(
 }
 
 /// Extends the rot rows to the full witness
+/// Input
+/// - witness: full witness of the circuit
+/// - word: 64-bit word to be rotated
+/// - rot:  rotation offset
+/// - side: side of the rotation, either left or right
+/// INTEGRATION:
```
[the remark this answers is not in the document]
Comment (querolita) [reply kimchi/src/circuits/polynomials/rot.rs:301]: ah well, assumed caps were more alerting 😅 but okay, I will change it for Warning

File: kimchi/src/tests/foreign_field_add.rs [hunk 12]
```diff
@@ -1453,3 +1456,106 @@ where
...
+    };
+
+    // witness
+    let witness = {
+        // create row for the public value 1
+        let mut witness: [_; COLUMNS] = array::from_fn(|_col| vec![Fp::zero(); 1]);
+        witness[0][0] = Fp::one();
+        // create inputs to the addition
+        let left = modulus.clone() - BigUint::from_bytes_be(&[1]);
+        let right = modulus.clone() - BigUint::from_bytes_be(&[1]);
```
[the remark this answers is not in the document]
Comment (querolita) [reply kimchi/src/tests/foreign_field_add.rs:1510]: Oh I had forgotten we had this, thanks

File: kimchi/src/tests/not.rs (hunk 4 above)
Comment (querolita) [reply kimchi/src/tests/not.rs:48]: Okay, I made the function inside the not polynomial to return an error, and then the function for tests is the one that panics

**Review by jspada — APPROVED** [review]

**Pull request merged**